File: NyaLauncher.Core/Download/ManifestGet.cs
using NyaLauncher.Core.Models;

namespace NyaLauncher.Core.Download;

/// <summary>
/// 获取 Minecraft 版本清单的服务。通过 <see cref="DownloadSourceProvider"/> 自动选择下载源。
/// </summary>
public static class ManifestGet
{
    /// <summary>
    /// 获取 Minecraft 版本清单（使用当前活跃下载源，失败自动回退）。
    /// </summary>
    public static async Task<List<MinecraftVersion>> GetVersionsAsync()
    {
        var json = await DownloadSourceProvider.GetStringAsync(
                DownloadSources.Official.LauncherMeta)
            .ConfigureAwait(false);

        var manifest = System.Text.Json.JsonSerializer.Deserialize<VersionManifest>(json);
        if (manifest is null)
            throw new System.Text.Json.JsonException("版本清单响应为空。");
        if (manifest.Latest is null)
            throw new System.Text.Json.JsonException("版本清单缺少 latest 节点。");
        if (manifest.Versions is null)
            throw new System.Text.Json.JsonException("版本清单缺少 versions 节点。");

        var versions = manifest.Versions.OfType<MinecraftVersion>().ToList();
        if (versions.Count == 0)
            return [];

        foreach (var version in versions)
        {
            if (version.Id == manifest.Latest.Release)
                version.IsLatestRelease = true;
            if (version.Id == manifest.Latest.Snapshot)
                version.IsLatestSnapshot = true;
        }

        return [.. versions.OrderByDescending(version => version.ReleaseTime)];
    }
}

File: internal/download/manifest_get.go
package download

import (
	"context"
	"encoding/json"
	"fmt"
	"sort"

	"nyalauncher/internal/models"
)

// ManifestGet 获取 Minecraft 版本清单的服务。通过 SourceProvider 自动选择下载源。
// 对应 C# ManifestGet。

// GetVersions 获取 Minecraft 版本清单（使用当前活跃下载源，失败自动回退）。
// 返回按发布时间降序的版本列表。
func GetVersions(ctx context.Context) ([]models.MinecraftVersion, error) {
	rawJSON, err := SourceProvider.GetString(ctx, DownloadSources.Official.LauncherMeta, nil)
	if err != nil {
		return nil, err
	}

	var manifest models.VersionManifest
	if err := json.Unmarshal([]byte(rawJSON), &manifest); err != nil {
		return nil, fmt.Errorf("版本清单响应为空或格式错误：%w", err)
	}
	versions := manifest.Versions
	if len(versions) == 0 {
		return []models.MinecraftVersion{}, nil
	}

	for i := range versions {
		if versions[i].ID == manifest.Latest.Release {
			versions[i].IsLatestRelease = true
		}
		if versions[i].ID == manifest.Latest.Snapshot {
			versions[i].IsLatestSnapshot = true
		}
	}

	sort.SliceStable(versions, func(i, j int) bool {
		return versions[i].ReleaseTime.After(versions[j].ReleaseTime)
	})
	return versions, nil
}
